internal/controller/promotionstrategy_controller.go

```diff
@@ -214,54 +211,54 @@ func (r *PromotionStrategyReconciler) upsertChangeTransferPolicy(ctx context.Con
-	for _, environment := range ps.Spec.Environments {
-		ctp, ok := ctpsByBranch[environment.Branch]
-		if !ok {
-			return fmt.Errorf("ChangeTransferPolicy not found for branch %s", environment.Branch)
-		}
-
-		ps.Status.Environments = utils.UpsertEnvironmentStatus(ps.Status.Environments, func() promoterv1alpha1.EnvironmentStatus {
-			status := promoterv1alpha1.EnvironmentStatus{
-				Branch: environment.Branch,
-				Active: promoterv1alpha1.PromotionStrategyBranchStateStatus{
-					Dry:      promoterv1alpha1.CommitShaState{Sha: ctp.Status.Active.Dry.Sha, CommitTime: ctp.Status.Active.Dry.CommitTime},
-					Hydrated: promoterv1alpha1.CommitShaState{Sha: ctp.Status.Active.Hydrated.Sha, CommitTime: ctp.Status.Active.Hydrated.CommitTime},
-					CommitStatus: promoterv1alpha1.PromotionStrategyCommitStatus{
-						Phase: string(promoterv1alpha1.CommitPhasePending),
-						Sha:   string(promoterv1alpha1.CommitPhasePending),
-					},
-				},
-				Proposed: promoterv1alpha1.PromotionStrategyBranchStateStatus{
-					Dry:      promoterv1alpha1.CommitShaState{Sha: ctp.Status.Proposed.Dry.Sha, CommitTime: ctp.Status.Proposed.Dry.CommitTime},
-					Hydrated: promoterv1alpha1.CommitShaState{Sha: ctp.Status.Proposed.Hydrated.Sha, CommitTime: ctp.Status.Proposed.Hydrated.CommitTime},
-					CommitStatus: promoterv1alpha1.PromotionStrategyCommitStatus{
-						Phase: string(promoterv1alpha1.CommitPhasePending),
-						Sha:   string(promoterv1alpha1.CommitPhasePending),
-					},
-				},
+// calculateStatus calculates the status of the PromotionStrategy based on the ChangeTransferPolicies.
+// ps.Spec.Environments must be the same length and in the same order as ctps.
+// This function updates ps.Status.Environments to be the same length and order as ps.Spec.Environments.
+func (r *PromotionStrategyReconciler) calculateStatus(ps *promoterv1alpha1.PromotionStrategy, ctps []*promoterv1alpha1.ChangeTransferPolicy) {
+	// Reconstruct current environment state based on ps.Environments order. Dropped environments will effectively be
+	// deleted, and new environments will be added as empty statuses. Those new environments will be populated in the
+	// ctp loop.
+	environmentStatuses := make([]promoterv1alpha1.EnvironmentStatus, len(ps.Spec.Environments))
+	for i, environment := range ps.Spec.Environments {
+		for _, environmentStatus := range ps.Status.Environments {
+			if environmentStatus.Branch == environment.Branch {
+				environmentStatuses[i] = environmentStatus
+				break
 			}
-
-			return status
-		}())
-
-		i, environmentStatus := utils.GetEnvironmentStatusByBranch(*ps, environment.Branch)
+		}
+	}
+	ps.Status.Environments = environmentStatuses
+
+	for i, ctp := range ctps {
+		// Update fields individually to avoid overwriting existing fields.
+		ps.Status.Environments[i].Branch = ctp.Spec.ActiveBranch
```

Comment:
As far as I could tell, there was no need to construct a new `CommitShaState`, since the struct only contains two fields anyway. So I just copy the one that's there.